syntax = "proto3";

package payloads.proto;

message ValueWrapper {

  oneof value {
    bool bool = 1;
    bytes bytes = 2;
    double double = 3;
    float float = 4;
    int32 int4 = 5;
    int64 int8 = 6;
    string string = 7;
    string json = 8;
  }

}

message RowResponse {

  repeated ValueWrapper values = 1;

}

message QuerySuccessResponse {

  repeated string column_names = 2;
  repeated RowResponse rows = 3;

}

message ErrorResponse {

  string message = 1;
  string correlation_id = 2;

  enum ErrorType {
    CONNECTION_CONFIG_ERROR = 0;
    CREATE_POOL_ERROR = 1;
    FORBIDDEN_CONNECTION_ID = 2;
    MISSING_CONNECTION = 3;
    POOL_ERROR = 4;
    POSTGRES_ERROR = 5;
    STORAGE_ERROR = 6;
    WRONG_NUM_OF_PARAMS = 7;
    UNKNOWN_PG_VALUE_TYPE = 8;
    LIMIT_EXCEEDED = 9;
  }

  ErrorType error_type = 3;
}

message QueryResponse {

  oneof payload {
    QuerySuccessResponse success = 1;
    ErrorResponse error = 2;
  }

}

message QueryRequest {

  string connection_id = 1;
  string query = 2;
  repeated ValueWrapper params = 3;

}

message TxBulkQueryParams {
  repeated ValueWrapper values = 1;
}

message TxBulkQueryRequestBody {

  string query = 1;
  repeated TxBulkQueryParams params = 2;

}

message TxBulkQueryRequest {

  string connection_id = 1;
  repeated TxBulkQueryRequestBody queries = 2;

}

message RowResponseGroup {

  repeated RowResponse rows = 1;

}

message TxQuerySuccessResponse {

  int32 ord_num = 1;
  repeated string column_names = 2;
  repeated RowResponseGroup row_groups = 3;

}

message TxBulkQuerySuccessResponse {

  repeated TxQuerySuccessResponse responses = 1;

}

message TxBulkQueryResponse {

  oneof payload {
    TxBulkQuerySuccessResponse success = 1;
    ErrorResponse error = 2;
  }

}


